### PR TITLE
Fix leaking proxy credentials

### PR DESCRIPTION
**Pull request by mdellweg**
fixes #9573

Should i create cherry-picks to the branches 3.8, 3.9 and 3.10 too?

### CODE REVIEW

**Comment by bmbouter** [comment]
The galaxy folks indicated at a sprint planning meeting that they will test this pre-merge and let us know. So given that I thought maybe putting this at draft would be the right thing to do.

**Review by bmbouter — APPROVED** [review]
This looks right to me, but let's wait for the final signoff from @rochacbruno. Also we need to hear what other pulpcore versions this is needed for, I'm not sure.

**Comment by ipanova** [comment]
@mdellweg regarding cherry-picks I think we should backport only to the requested branch(es).

**Comment by rochacbruno** [comment]
Hi @mdellweg I tested with this patch and did not solved the problem as the call is being done not on `factory.py` but in `http.py`

this is where the proxy credentials needs to be sanitized,

```py
 File "/src/pulpcore/pulpcore/download/http.py", line 209, in _run
worker_1            |     async with self.session.get(self.url, proxy=self.proxy, auth=self.auth) as response:
``` 

Full traceback

```py
worker_1            | Giving up _run(...) after 1 tries (aiohttp.client_exceptions.ClientHttpProxyError: 400, message='Bad Request', url=URL('http://foo:bazinga@46.4.96.137:3128'))
worker_1            | pulp: backoff:ERROR: Giving up _run(...) after 1 tries (aiohttp.client_exceptions.ClientHttpProxyError: 400, message='Bad Request', url=URL('http://foo:bazinga@46.4.96.137:3128'))
worker_1            | Giving up _run(...) after 1 tries (aiohttp.client_exceptions.ClientHttpProxyError: 400, message='Bad Request', url=URL('http://foo:bazinga@46.4.96.137:3128'))
worker_1            | pulp: backoff:ERROR: Giving up _run(...) after 1 tries (aiohttp.client_exceptions.ClientHttpProxyError: 400, message='Bad Request', url=URL('http://foo:bazinga@46.4.96.137:3128'))
worker_1            | pulp: rq.worker:ERROR: Traceback (most recent call last):
worker_1            |   File "/venv/lib64/python3.6/site-packages/rq/worker.py", line 936, in perform_job
worker_1            |     rv = job.perform()
worker_1            |   File "/venv/lib64/python3.6/site-packages/rq/job.py", line 684, in perform
worker_1            |     self._result = self._execute()
worker_1            |   File "/venv/lib64/python3.6/site-packages/rq/job.py", line 690, in _execute
worker_1            |     return self.func(*self.args, **self.kwargs)
worker_1            |   File "/venv/lib64/python3.6/site-packages/pulp_ansible/app/tasks/collections.py", line 83, in sync
worker_1            |     d_version.create()
worker_1            |   File "/src/pulpcore/pulpcore/plugin/stages/declarative_version.py", line 148, in create
worker_1            |     loop.run_until_complete(pipeline)
worker_1            |   File "/usr/lib64/python3.6/asyncio/base_events.py", line 484, in run_until_complete
worker_1            |     return future.result()
worker_1            |   File "/src/pulpcore/pulpcore/plugin/stages/api.py", line 225, in create_pipeline
worker_1            |     await asyncio.gather(*futures)
worker_1            |   File "/src/pulpcore/pulpcore/plugin/stages/api.py", line 43, in __call__
worker_1            |     await self.run()
worker_1            |   File "/venv/lib64/python3.6/site-packages/pulp_ansible/app/tasks/collections.py", line 313, in run
worker_1            |     async for metadata in self._fetch_collections():
worker_1            |   File "/venv/lib64/python3.6/site-packages/pulp_ansible/app/tasks/collections.py", line 454, in _fetch_collections
worker_1            |     count = await _loop_through_pages(not_done)
worker_1            |   File "/venv/lib64/python3.6/site-packages/pulp_ansible/app/tasks/collections.py", line 419, in _loop_through_pages
worker_1            |     url = await _get_url(1, versions_url)
worker_1            |   File "/venv/lib64/python3.6/site-packages/pulp_ansible/app/tasks/collections.py", line 402, in _get_url
worker_1            |     api_endpoint = (await _get_collection_api(root))[0]
worker_1            |   File "/venv/lib64/python3.6/site-packages/pulp_ansible/app/tasks/collections.py", line 374, in _get_collection_api
worker_1            |     api_data = parse_metadata(await downloader.run())
worker_1            |   File "/src/pulpcore/pulpcore/download/base.py", line 227, in run
worker_1            |     return await self._run(extra_data=extra_data)
worker_1            |   File "/venv/lib64/python3.6/site-packages/backoff/_async.py", line 133, in retry
worker_1            |     ret = await target(*args, **kwargs)
worker_1            |   File "/venv/lib64/python3.6/site-packages/pulp_ansible/app/downloaders.py", line 88, in _run
worker_1            |     return await super()._run(extra_data=extra_data)
worker_1            |   File "/venv/lib64/python3.6/site-packages/backoff/_async.py", line 133, in retry
worker_1            |     ret = await target(*args, **kwargs)
worker_1            |   File "/src/pulpcore/pulpcore/download/http.py", line 209, in _run
worker_1            |     async with self.session.get(self.url, proxy=self.proxy, auth=self.auth) as response:
worker_1            |   File "/venv/lib64/python3.6/site-packages/aiohttp/client.py", line 1117, in __aenter__
worker_1            |     self._resp = await self._coro
worker_1            |   File "/venv/lib64/python3.6/site-packages/aiohttp/client.py", line 521, in _request
worker_1            |     req, traces=traces, timeout=real_timeout
worker_1            |   File "/venv/lib64/python3.6/site-packages/aiohttp/connector.py", line 535, in connect
worker_1            |     proto = await self._create_connection(req, traces, timeout)
worker_1            |   File "/venv/lib64/python3.6/site-packages/aiohttp/connector.py", line 890, in _create_connection
worker_1            |     _, proto = await self._create_proxy_connection(req, traces, timeout)
worker_1            |   File "/venv/lib64/python3.6/site-packages/aiohttp/connector.py", line 1129, in _create_proxy_connection
worker_1            |     headers=resp.headers,
worker_1            | aiohttp.client_exceptions.ClientHttpProxyError: 400, message='Bad Request', url=URL('http://foo:bazinga@46.4.96.137:3128')

```

**Comment by rochacbruno** [comment]
I just confirmed the above statements

```py
worker_1            |   File "/src/pulpcore/pulpcore/download/http.py", line 227, in _run
worker_1            |     auth=self.auth
worker_1            |   File "/venv/lib64/python3.6/site-packages/aiohttp/client.py", line 1117, in __aenter__
worker_1            |     self._resp = await self._coro
worker_1            |   File "/venv/lib64/python3.6/site-packages/aiohttp/client.py", line 521, in _request
worker_1            |     req, traces=traces, timeout=real_timeout
worker_1            |   File "/venv/lib64/python3.6/site-packages/aiohttp/connector.py", line 535, in connect
worker_1            |     proto = await self._create_connection(req, traces, timeout)
worker_1            |   File "/venv/lib64/python3.6/site-packages/aiohttp/connector.py", line 890, in _create_connection
worker_1            |     _, proto = await self._create_proxy_connection(req, traces, timeout)
worker_1            |   File "/venv/lib64/python3.6/site-packages/aiohttp/connector.py", line 1129, in _create_proxy_connection
worker_1            |     headers=resp.headers,
worker_1            | aiohttp.client_exceptions.ClientHttpProxyError: 400, message='Bad Request', url=URL('http://46.4.96.137:3128')

```

I used 

```patch
diff --git a/pulpcore/download/http.py b/pulpcore/download/http.py
index c8919af43..bbdafae5b 100644
--- a/pulpcore/download/http.py
+++ b/pulpcore/download/http.py
@@ -3,6 +3,7 @@ import logging
 import aiohttp
 import backoff
 
+from urllib.parse import urlparse, urlunparse
 from .base import BaseDownloader, DownloadResult
 
 
@@ -206,7 +207,25 @@ class HttpDownloader(BaseDownloader):
         Args:
             extra_data (dict): Extra data passed by the downloader.
         """
-        async with self.session.get(self.url, proxy=self.proxy, auth=self.auth) as response:
+        proxy_url = self.proxy
+        proxy_auth = None
+        if proxy_url:
+            parsed_url = urlparse(proxy_url)
+            netloc = parsed_url.netloc
+            if "@" in netloc:
+                auth, url = netloc.rsplit("@", maxsplit=1)
+                proxy_username, proxy_password = auth.split(":", maxsplit=1)
+                proxy_url = urlunparse(parsed_url._replace(netloc=url))
+                proxy_auth = aiohttp.BasicAuth(
+                    login=proxy_username, password=proxy_password
+                )
+
+        async with self.session.get(
+            self.url,
+            proxy=proxy_url,
+            proxy_auth=proxy_auth,
+            auth=self.auth
+        ) as response:
             self.raise_for_status(response)
             to_return = await self._handle_response(response)
             await response.release()

```

**Comment by ironfroggy** [comment]
@bmbouter @rochacbruno I've tested this PR against an install of galaxy_ng 4.2 and still get the password in errors. I looked at the change and I agree it _looks_ like it could work, but I've confirmed I have the branch installed on my docker image and the issue is definitely still present. Let me know if I can provide any further information that might help.

**Comment by mdellweg** [comment]
> I used

Thank you very much for the analysis @rochacbruno !
Do you agree that we should fix it in both places?

**Comment by mdellweg** [comment]
@rochacbruno I implemented in a slightly different way. (Looks like the `proxy_auth` parameter to the HttpDownloader was basically ignored before.)

Do you mind re-testing?

**Comment by rochacbruno** [comment]
@mdellweg I just tested and it works, this is now showing only the URL error and no password or username

```bash
worker_1            |   File "/venv/lib64/python3.6/site-packages/aiohttp/client_reqrep.py", line 366, in update_host
worker_1            |     raise InvalidURL(url)
worker_1            | aiohttp.client_exceptions.InvalidURL: 46.4.96.137:3128
```
Screenshot_2021-11-29_16-45-37

**Review by bmbouter — APPROVED** [review]
Thanks!

**Review by dkliban — APPROVED** [review]

**Comment by mdellweg** [comment]
Looks like the docs branch suffers from the rq/click incompatibility.

**Comment by mdellweg** [comment]
somehow `functest_requirements` was allowed to install `click~=8` over the explicitely installed `click<8` from `docs_requirements`.
This package manager... :imp:

File: pulpcore/download/http.py [hunk 1]
```diff
@@ -206,7 +220,9 @@ async def _run(self, extra_data=None):
         Args:
             extra_data (dict): Extra data passed by the downloader.
         """
-        async with self.session.get(self.url, proxy=self.proxy, auth=self.auth) as response:
+        async with self.session.get(
```
Comment (ipanova) [inline pulpcore/download/http.py:223]: @rochacbruno is this fix needed only for galaxy? some plugins subclass the downloaders, like pulp_container, and this line would need to be adjusted there as well.

Comment (rochacbruno) [reply pulpcore/download/http.py:223]: Hi @ipanova on pulp_ansible this is already fixed, I am not sure about others but I guess any plugin that makes http connection can be behind a proxy, so pulp_container might be affected, but this specific issue is for pulpcore 3.7 so galaxy was not using pulp_container yet I guess.

Comment (ipanova) [reply pulpcore/download/http.py:223]: great,thanks,  it answers my question then - otherwise we would have needed to also release pulp_container.

**Pull request merged**